pennylane/transforms/qcut.py

```diff
@@ -505,3 +505,160 @@ def contract_tensors(
     kwargs = {} if use_opt_einsum else {"like": tensors[0]}
 
     return contract(eqn, *tensors, **kwargs)
+
+
+CHANGE_OF_BASIS = qml.math.array([[1.0, 1, 0, 0], [-1, -1, 2, 0], [-1, -1, 0, 2], [1, -1, 0, 0]])
+
+
+def _process_tensor(results, n_prep: int, n_meas: int):
+    """Convert a flat slice of execution results into a tensor.
+
+    This function performs the following steps:
+
+    1. Reshapes ``results`` into the shape ``(4,) * n_prep + (4**n_meas,)``
```

Comment:
Unfortunately a lot of this PR seems to hinge on understanding the manipulation of these tensor dimensions which is a blocker for me at the moment. 

Will keep thinking about this :thinking:

Reply:
Hey @anthayes92!

> Or perhaps it's (4,4,4) + (16), = (4, 4, 4, 16)

Exactly, if `n_prep = 3` and `n_meas = 2` then we ideally want `(4, 4, 4, 4, 4)`.

We start with a flat list of size `4 ** 5 = 1024`. We reshape to `(4, 4, 4, 16)` so that we have the right indices for the preparations (first 3 indices) but keep things flat for the measurements (the 16) index. We keep things flat for the measurements because we need to shuffle from:
`II, IZ, ZI, ZZ, IX, ZX, IY, ZY, XI, XZ, XX, XY, YI, YZ, YX, YY` (of length 16) to:
`II, IX, IY, IZ, XI, XX, XY, ..., ZZ` (of length 16).
Once we've shuffled, we reshape to `(4, 4, 4, 4, 4)`, and then do a change of basis on the first 3 indices.

> Will keep thinking about this

Sounds good, let me know if I can help with any more understanding!

Reply:
Ah this makes thanks, all indices accounted for!